app/controllers/turbo/requests.rb

```diff
@@ -9,7 +9,21 @@
 # Turbo Frames knows how to fish out the relevant frame regardless.
 #
 # This module is automatically included in <tt>ActionController::Base</tt>.
-module Turbo::Frames::FrameRequest
+#
+# You can override this behavior in <tt>ApplicationController</tt> if you want to have a layout for turbo stream requests:
+#
+#   # app/controllers/articles_controller.rb
+#   class ApplicationController < ActionController::Base
+#     layout -> { (turbo_frame_request? && !turbo_stream_request?) ? false : "application" }
```

Comment:
In most cases, Turbo will submit form requests with the Turbo Stream `Accept:` header present, so `turbo_stream_request?` would evaluate to `true`, and any response would be rendered with the `"application"` layout.

Unfortunately, whether or not a response uses a `.html.erb` or `.turbo_stream.erb` template hinges on the **response**, and not the **request**.

For example, consider a `POST` request to `/messages` with an `Accept:` header containing the Turbo Stream MIME type, and a `Turbo-Frame: my-frame` header. The presence of both headers means that:

* `turbo_frame_request?` evaluates to `true`
* `turbo_stream_request?` evaluates to `true`
* `!turbo_stream_request?` evaluates to `false`
* `turbo_frame_request? && !turbo_stream_request?` evaluates to `false`

That means that the `layout` expression would return `"application"`.

Consider a hypothetical `messages#create` controller action that supports both HTML and Turbo Stream responses:

```ruby
def create
  @message = Message.new(messages_params)

  if @message.save
    respond_to do |format|
      format.html { redirect_to @message, status: :see_other }
      format.turbo_stream
    end
  else
    render :new, status: :unprocessable_entity
  end
end
```

In the case of the `format.turbo_stream` block, calls to `turbo_stream_request?` would return `true`, and the documentation's guidance makes sense: render the `messages/create.turbo_stream.erb` template into the `app/views/layouts/application.turbo_stream.erb` layout.

However, since the `POST` request's `Accept:` header includes the Turbo Stream MIME type, calls to `turbo_stream_request?` in the `format.html` branch would also evaluate to `true`. If the subsequent request that follows the redirect is made with the same header value (`Turbo-Frame` and `Accept`, for example) as the original `POST` request, then calls to `turbo_stream_request?` and `turbo_frame_request?` within that request would also evaluate to `true`. This means that no matter what, responses are rendered into an `"application"` layout.

Now consider a `messages#create` controller action that only responds with HTML:

```diff
 def create
   @message = Message.new(messages_params)
 
   if @message.save
-    respond_to do |format|
-      format.html { redirect_to @message, status: :see_other }
-      format.turbo_stream
-    end
+    redirect_to @message
   else
     render :new, status: :unprocessable_entity
   end
 end
```

The absence of a [respond_to](https://edgeapi.rubyonrails.org/classes/ActionController/MimeResponds.html#method-i-respond_to) call combined with the `redirect_to @message` means that the response is going to have an HTML MIME type, not a Turbo Stream MIME type. In that scenario, the `layout` block's inclusion of `!turbo_stream_request?` would still cause the response body to render with the `application` layout.

Personally, I'm in favor of [always rendering responses into a layout](https://github.com/hotwired/turbo-rails/pull/232), but that would require a broader discussion than this one.

Am I misunderstanding something about this process, or something about your particular use case that makes this scenario a non-issue?

Reply:
Thanks again for your response. I agree with everything you said above, and you are right for the `redirect_to @message`, I forgot to consider this case as I rarely use it.

The use case was consider a `MessagesController#new` action: 

```rb
def new
  @message = Message.new
end
```

If the link to the new message page is in a Turbo Frame, we would have the following headers: 

```txt
Accept: text/html, application/xhtml+xml
Turbo-Frame: new_article
```

This means that `layout -> { (turbo_frame_request? && !turbo_stream_request?) ? false : "application" }` would evaluate to `false` and so in that specific case, we wouldn't have a layout.

I will close the PR as it does not work in all cases. Thanks again for your review!

Reply:
I think the answer is that we need a way at a top-level to set whether you want to use layouts for turbo frames (rendering them in layouts/application.turbo_frame.html) and then document how to overwrite that on a per-action basis with `format.turbo_frame { render layout: "something_else" }`.